### DISCUSSION TITLE
New findings in the Viomi map format

### DISCUSSION THREAD

**Discussion by halfbakery** [Hypfer/Valetudo#966] [Archive]
I'm a Viomi V2 Pro (viomi.vacuum.v6) owner, and I created a Viomi map parser for https://github.com/PiotrMachowski/Home-Assistant-custom-components-Xiaomi-Cloud-Map-Extractor/tree/dev_new_api based on your excellent documentation. During the implementation I've found some new information which I'd like to share:

- The map definetly has a structure: as you realized, the map starts with the `featureFlags` word or dword, followed by the `mapId` field. It seems the `mapId` value is the section header magic. I don't know whether it's version specific or random (my findings are based on just some maps, but I've never ever seen the same value). All of the time I've found the `featureFlags` 0x78ff and the map exactly has 12 sections. The `mapId` at offset 4 is the magic of the robot status section.
- map section: these are the values I've encountered:
  - `0` / `255`: the same as you found
  - `1`: scanned but unreachable area
  - `2`: I couldn't find its meaning, but found on many maps
  - `[10 ... 59]`: room_id
  - `[60 ... 109]`: rooms selected/under cleanup, `room_id = value - 50`
- history section: the first byte is taxiing (0x00) or working (0x01). I've tested it in both vacuuming or vacuuming+mopping mode.
- room section: in each room entry (id, name, unknown) the unknown 8 bytes is a coordinate pair (two float32) of the room center/room name position on the map.

**Comment by Hypfer** [comment]
Great!

It would be awesome if you could update the kaitai struct with your findings:
https://github.com/Hypfer/Valetudo/blob/master/assets/kaitai_structs/viomi_map_file.ksy

For testing, just throw that + a sample map file into https://ide.kaitai.io/
You should then get a nice interactive dissection
